crates/cranelift/src/compiler/component.rs

```diff
@@ -262,7 +328,337 @@ impl<'a> TrampolineCompiler<'a> {
                 let params = self.builder.func.dfg.block_params(self.block0);
                 (params[2], params[3])
             }
-        };
+        }
+    }
+
+    fn translate_intrinsic_libcall(
+        &mut self,
+        vmctx: ir::Value,
+        get_libcall: fn(
+            &dyn TargetIsa,
+            &mut ir::Function,
+        ) -> (ir::SigRef, ComponentBuiltinFunctionIndex),
+        args: &[ir::Value],
+        result: ir::types::Type,
+    ) {
+        match self.abi {
+            Abi::Wasm => {}
+
+            Abi::Array => {
+                // TODO: A guest could hypothetically export the same intrinsic
+                // it imported, allowing the host to call it directly.  We need
+                // to support that here (except for `sync-enter`, `sync-exit`,
+                // `async-enter`, and `async-exit`, which are only ever called
+                // from FACT-generated Wasm code and never exported).
+                self.builder.ins().trap(TRAP_INTERNAL_ASSERT);
+                return;
+            }
+        }
+
+        let call = self.call_libcall(vmctx, get_libcall, args);
+
+        if result == ir::types::I64 {
+            let result = self.builder.func.dfg.inst_results(call)[0];
+            let result = self.raise_if_negative_one(result);
+            self.abi_store_results(&[result]);
+        } else {
+            if result != ir::types::I8 {
+                todo!("support additional intrinsic return types")
+            }
+            let succeeded = self.builder.func.dfg.inst_results(call)[0];
+            self.raise_if_host_trapped(succeeded);
+            self.builder.ins().return_(&[]);
+        }
+    }
+
+    fn translate_task_return_call(&mut self, results: TypeTupleIndex) {
+        let args = self.builder.func.dfg.block_params(self.block0).to_vec();
+        let vmctx = args[0];
+
+        let (values_vec_ptr, values_vec_len) = self.store_wasm_arguments(&args[2..]);
+
+        let ty = self
+            .builder
+            .ins()
+            .iconst(ir::types::I32, i64::from(results.as_u32()));
+
+        self.translate_intrinsic_libcall(
+            vmctx,
+            host::task_return,
+            &[vmctx, ty, values_vec_ptr, values_vec_len],
+            ir::types::I8,
+        );
+    }
+
+    fn translate_sync_enter(&mut self) {
+        match self.abi {
+            Abi::Wasm => {}
+
+            Abi::Array => {
+                // This code can only be called from (FACT-generated) Wasm, so
+                // we don't need to support the array ABI.
+                self.builder.ins().trap(TRAP_INTERNAL_ASSERT);
+                return;
+            }
+        }
+
+        let args = self.builder.func.dfg.block_params(self.block0).to_vec();
+        let vmctx = args[0];
+
+        let pointer_type = self.isa.pointer_type();
+        let wasm_func_ty = &self.types[self.signature].unwrap_func();
+
+        let param_offset = 5;
+        let spill_offset = param_offset + 2;
```

Comment:
If you can comment pointing to a location in the fact trampoline bits to where these constants come from (and vice versa ideally) that'd be good to keep the two in sync. (and/or extracting this to constants or similar)